xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+        for ( int t = start; t < end; t++ ) {
+            #ifdef DEBUG_MEASURE_TEXT
+                printf("%sMSHB t%d (=%x) ", indent, t, m_text[t]);
+            #endif
+            // Grab all glyphs that do not belong to a cluster greater that our char position
+            int cur_width = 0; // current cluster width
+            while ( hg < glyph_count ) {
+                hcl = glyph_info[hg].cluster;
+                if ( hcl <= t ) { // glyph still part of a previous cluster
+                    int advance = 0;
+                    if ( glyph_info[hg].codepoint != 0 ) { // Codepoint found in this font
+                        #ifdef DEBUG_MEASURE_TEXT
+                            printf("(found cp=%x) ", glyph_info[hg].codepoint);
+                        #endif
+                        if ( t_notdef_start >= 0 ) { // But we have a segment of previous ".notdef"
+                            t_notdef_end = t;
+
+                            // Let a fallback font replace the wrong values in widths and flags
+                            // No-op if there is no more fallback font
+                            #ifdef DEBUG_MEASURE_TEXT
+                                printf("%s[...]\n%sMSHB ### measuring past failures with fallback font %d>%d\n",
+                                                        indent, indent, t_notdef_start, t_notdef_end);
+                            #endif
+                            // Drop BOT/EOT flags if this segment is not at start/end
+                            int fb_hints = hints;
+                            if ( t_notdef_start > 0 )
+                                fb_hints &= ~HINT_BEGINS_PARAGRAPH;
+                            if ( t_notdef_end < len )
+                                fb_hints &= ~HINT_ENDS_PARAGRAPH;
+                            int fallback_width = measureSegment( font_num+1, t_notdef_start, t_notdef_end, fb_hints );
+                            if ( fallback_width != NOT_MEASURED ) {
+                                // The individual char widths will have been updated,
+                                // but we need to correct final_width where we kept
+                                // adding notdef widths
+                                final_width = final_width - notdef_width + fallback_width;
+                            }
+                            #ifdef DEBUG_MEASURE_TEXT
+                                printf("%sMSHB ### measured past failures > W= %d\n%s[...]",
+                                                                indent, fallback_width, indent);
+                            #endif
+
+                            t_notdef_start = -1;
+                            notdef_width = 0;
+                            // And go on with the found glyph now that we fixed what was before
+                        }
+                        // Glyph found in this font
+                        advance = FONT_METRIC_TRUNC(FONT_METRIC_ROUND(glyph_pos[hg].x_advance));
```

Comment:
I'm not *quite* sure why you need to TRUNC?

ROUND does make sense, AFAIU (which, granted, isn't much ;D).

Reply:
Okay, see `shapeSegments` ;).